pkg/applicationserver/applicationserver.go

```diff
@@ -1486,6 +1538,15 @@ func (as *ApplicationServer) decryptDownlinkMessage(ctx context.Context, ids *tt
 	if err != nil {
 		return err
 	}
+
+	entity, err := as.endDeviceRegistry.Get(ctx, ids, []string{"attributes"})
```

Comment:
Now you have a sequential set and get. On set, you can pass `attributes` to the field mask argument of `Set()`. That makes it available in the callback so you can put it in `msg.Attributes`. Don't return `attributes` from the callback (like now) because otherwise it would be actually set. Then you can remove this `Get()` entirely.

Reply:
Do you mean doing something like this?

```go
if _, err = as.endDeviceRegistry.Set(ctx, ids, []string{"attributes"},
	func(entity *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
		if entity == nil {
			return nil, nil, errDeviceNotFound.WithAttributes("device_uid", unique.ID(ctx, ids))
		}
		msg.Attributes = entity.Attributes
		return entity, []string{""}, nil
}); err != nil {
	log.FromContext(ctx).WithError(err).Warn(
		"Failed to retrieve end device attributes on downlink nack",
        )
}
```

Reply:
No, I'm referring to the `Set()` above that sets `location`.

Reply:
Oh, but that one is in another function

Reply:
Ah I think I overlooked that, thanks.